tests/undefined_symbol_in_sdfg_test.py

```diff
@@ -0,0 +1,344 @@
+    sdfg.add_array('B', [20], dace.float64)
+    sdfg.add_transient('tmp', [20, undefined_dim], dace.float64)
+
+    # Create state
+    state = sdfg.add_state()
+
+    # Create access nodes
+    A = state.add_read('A')
+    tmp = state.add_access('tmp')
+    B = state.add_write('B')
+
+    # Add map
+    map_entry, map_exit = state.add_map('compute', dict(i='0:20', j='0:10'))
+
+    # Add tasklet
+    tasklet = state.add_tasklet('compute', {'a'}, {'b', 't'}, 'b = a; t = a')
+
+    # Connect with edges that use the undefined symbol
+    state.add_edge(A, None, map_entry, None, dace.Memlet())
+    state.add_edge(map_entry, None, tasklet, None, dace.Memlet())
+    state.add_edge(A, None, tasklet, 'a', dace.Memlet('A[i, j]'))
+    state.add_edge(tasklet, 't', tmp, None, dace.Memlet('tmp[i, j]'))
+    state.add_edge(tasklet, 'b', B, None, dace.Memlet('B[i]'))
+    state.add_edge(tasklet, None, map_exit, None, dace.Memlet())
+
+    # This should fail validation because the transient has an undefined dimension
+    with pytest.raises(InvalidSDFGError, match="undefined symbol in dimension"):
+        sdfg.validate()
+
+
+def test_undefined_symbol_value_assignment():
+    """Tests that an UndefinedSymbol can be assigned a value before validation to avoid errors."""
+
+    # Create an SDFG manually
+    sdfg = dace.SDFG('undefined_symbol_assignment_test')
+
+    # Define symbols
+    undefined_dim = symbolic.UndefinedSymbol()
+
+    # Add arrays to SDFG - use the undefined symbol in the shape
+    sdfg.add_array('A', [20, 20], dace.float64)
+    sdfg.add_array('B', [20], dace.float64)
+    sdfg.add_transient('tmp', [20, undefined_dim], dace.float64)
+
+    # Create state
+    state = sdfg.add_state()
+
+    # Create access nodes
+    A = state.add_read('A')
+    tmp = state.add_access('tmp')
+    B = state.add_write('B')
+
+    # Add map
+    map_entry, map_exit = state.add_map('compute', dict(i='0:20', j='0:10'))
+
+    # Add tasklet
+    tasklet = state.add_tasklet('compute', {'a'}, {'b', 't'}, 'b = a; t = a')
+
+    # Connect with edges that use the undefined symbol
+    state.add_edge(A, None, map_entry, None, dace.Memlet())
+    state.add_edge(map_entry, None, tasklet, None, dace.Memlet())
+    state.add_edge(A, None, tasklet, 'a', dace.Memlet('A[i, j]'))
+    state.add_edge(tasklet, 't', tmp, None, dace.Memlet('tmp[i, j]'))
+    state.add_edge(tasklet, 'b', B, None, dace.Memlet('B[i]'))
+    state.add_edge(tasklet, None, map_exit, None, dace.Memlet())
+
+    # Replace undefined symbols in the SDFG with concrete values
+    # This shows how a user would explicitly replace UndefinedSymbol with concrete values
+    replaced_shape = []
+    for dim in sdfg.arrays['tmp'].shape:
+        if isinstance(dim, symbolic.UndefinedSymbol):
+            replaced_shape.append(10)
+        else:
+            replaced_shape.append(dim)
+    sdfg.arrays['tmp'].shape = tuple(replaced_shape)
+
+    # Make sure we can now validate the SDFG
+    sdfg.validate()
```

Comment:
I think this is a great test that should actually fail validation. As mentioned above, since this is a transient array, undefined symbols are not allowed. I would add a `with pytest.raises: sdfg.validate()` above the `replaced_shape` part, and another validation pass here. Note that based on my above comment, this SDFG should still fail validation. The reason for that is that the strides should be `[?, 1]` and total size should be `?`. I would test for this failure mode here by adding another `with pytest.raises`.